File: euler_082.py
from time import time
import numpy as np

matrix = np.empty([80, 80], int)

start = time()

with open("p082_matrix.txt", "r") as f:
    lines = f.readlines()
    for i in range(len(lines)):
        line = list(map(int, lines[i].replace("\n", "").split(",")))
        for j in range(len(line)):
            matrix[i, j] = line[j]

minimal = 300000

#If the optimal route ends in (i0, 79)
for i0 in range(80):
    path_sums = np.empty([80, 80], int)

    #Last column is initialised
    path_sums[i0, 79] = matrix[i0, 79]

    for i in range(1, i0 + 1):
        path_sums[i0 - i, 79] = matrix[i0 - i, 79] + path_sums[i0 - i + 1, 79]

    for i in range(i0 + 1, 80):
        path_sums[i, 79] = matrix[i, 79] + path_sums[i - 1, 79]

    #The penultimate row is populated
    path_sums[i0, 78] = matrix[i0, 78] + path_sums[i0, 79]

    for i in [k for k in range(80) if k != i0]:
        if i < i0:
            path_sums[i, 78] = matrix[i, 78] + min(path_sums[i, 79], path_sums[i + 1, 78])
        if i > i0:
            path_sums[i, 78] = matrix[i, 78] + min(path_sums[i - 1, 78], path_sums[i, 79])


    #The rest of the rows are populated
    for j in reversed(range(78)):
        for i in range(80):
            #go right
            optimal = matrix[i, j] + path_sums[i, j + 1]
            #go up
            if i > 0:
                k = 1
                while i - k > -1:
                    test = sum([matrix[i - d, j] for d in range(k + 1)]) + path_sums[i - k, j + 1]
                    if test < optimal:
                        optimal = test
                    k += 1
            #go down
            if i < 79:
                k = 1
                while i + k < 80:
                    test = sum([matrix[i + d, j] for d in range(k + 1)]) + path_sums[i + k, j + 1]
                    if test < optimal:
                        optimal = test
                    k += 1

            path_sums[i, j] = optimal

    minimal = min(minimal, min(path_sums[:, 0]))


print("Minimal path sum: " + str(minimal))

stop = time()
print("Time: " + str(stop - start))

File: euler_082.js
const start = Date.now();

import { readFileSync } from 'fs';

function syncReadFile(filename) {
    const contents = readFileSync(filename, 'utf-8');

    const arr = contents.split("\n");

    return arr.map(row => row.split(",").map(cell => Number(cell)));
}

const matrix2D = syncReadFile('./euler_082.txt');

// From row x col (0..79 x 0..79) to index in array (0..6399)
const index = (row, col) => {
    return row * 80 + col;
}

// Converting 2-dim matrix to 1-dim array
let matrix = Array(6400);
for (let row = 0; row < 80; row++) {
    for (let col = 0; col < 80; col++) {
        matrix[index(row, col)] = matrix2D[row][col];
    }
}
const entry = (row, col) => {
    return matrix[index(row, col)];
}

// The sub-sum matrix
let SSM = Array(6400);
// Last column is the same as the matrix
for (let i = 0; i < 80; i++) {
    SSM[index(i, 79)] = entry(i, 79);
}

for (let c = 78; c >= 0; c--) {
    for (let r = 0; r < 80; r++) {
        // Straight right
        let minVal = entry(r,c) + SSM[index(r, c + 1)];
        let exceededMin = false;
        let current = entry(r,c);
        // Look up
        for (let rl = r - 1; !exceededMin && rl >= 0; rl--) {
            current += entry(rl, c);
            exceededMin = current >= minVal;
            minVal = Math.min(minVal, current + SSM[index(rl, c + 1)]);
        }
        // Look down
        exceededMin = false;
        current = entry(r,c);
        for (let rl = r + 1; !exceededMin && rl < 80; rl++) {
            current += entry(rl, c);
            exceededMin = current >= minVal;
            minVal = Math.min(minVal, current + SSM[index(rl, c + 1)]);
        }
        SSM[index(r,c)] = minVal;
    }
}


let firstCol = [];
for (let i = 0; i < 80; i++) {
    firstCol.push(SSM[index(i, 0)]);
}

const minSum = Math.min(...firstCol);

const end = Date.now();

console.log("Execute time: ", end - start, " Minimal path sum is: ", minSum);